Added sentiment, syntax and entity analysis methods

shivanighanta opened the pull request.

shivanighanta committed:
Added sentiment, syntax and entity analysis methods

jbiral commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -0,0 +1,325 @@
+  public float getSentimentScore(String message) throws Exception {
+    Document doc =
+        Document.newBuilder().setContent(message).setType(Document.Type.PLAIN_TEXT).build();
+    LanguageServiceClient languageService = LanguageServiceClient.create();
```
Could you use a try with resources here as well?

jbiral commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  /** Identifies entities in a given string. */
+  public static void analyzeEntitiesText(String text) throws Exception {
```
I would recommend not making these static, so you can share class variables if needed.

jbiral reviewed (changes requested):
Overall, this is looking good!

A couple things I'd like to ask / point out:
1. I'd like to see how `doGet` is instantiating and calling the `Analyze` class to have more clarity on its intended use. It's fine to call it with dummy data for now and include it in the PR. This way, Bless can update the dummy data later to real data once it's ready.
1. Do we need to close the Service after each request, or can we re-use it multiple times before closing it? If it's the latter, I would suggest to keep the service open for the duration of the `doGet` method call, so we don't need to re-create the service if we have multiple posts to process.
1. Try to sub out the service, as suggested in the pod chat, so tests can be kept while not calling the real endpoint.
1. The lint / formatting checks are broken, so make sure to address them as otherwise the PR cannot be merged.

shivanighanta committed:
merge
Merge branch 'master' of https://github.com/googleinterns/step64-2020 into Sentiment-Analysis

ginellegaisano commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -0,0 +1,325 @@
+    LanguageServiceClient languageService = LanguageServiceClient.create();
+    Sentiment sentiment = languageService.analyzeSentiment(doc).getDocumentSentiment();
+    float score = sentiment.getScore();
+    // float score = Math.round(sentiment.getScore() * 100.0) / 100.0;
```
Nit: delete commented line.

ginellegaisano commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -0,0 +1,325 @@
+    /** Run sentiment analysis and return sentiment score */
+  public float getSentimentScore(String message) throws Exception {
+    Document doc =
+        Document.newBuilder().setContent(message).setType(Document.Type.PLAIN_TEXT).build();
```
If this, line 63, line 125 need to be kept in-sync, it might be worth creating a helper method to share between.

ginellegaisano commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -0,0 +1,325 @@
+      AnalyzeEntitiesResponse response = language.analyzeEntities(request);
+
+      // Print the response
+      for (Entity entity : response.getEntitiesList()) {
```
If you plan on keeping this, extract code into separate method to use on line 105 as well.

ginellegaisano commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -0,0 +1,325 @@
+      Document doc = Document.newBuilder().setContent(text).setType(Type.PLAIN_TEXT).build();
+      AnalyzeSentimentResponse response = language.analyzeSentiment(doc);
+      Sentiment sentiment = response.getDocumentSentiment();
+      if (sentiment == null) {
```
Same here - extract into helper to share with #analyzeSentimentFile

ginellegaisano reviewed (commented):
I left a few comments - generally looks good, though would recommend going through and refactoring so that methods with duplication can share code. I noticed that the difference between a lot of these pairs of functions is that one sets a gcsUri and the other sets text. The rest can go into a helper. (ie. analyzeEntitiesFile and analyzeEntitiesText can have a helper called analyzeEntities(Document doc)).

shivanighanta commented:
> Overall, this is looking good!
> 
> A couple things I'd like to ask / point out:
> 
> 1. I'd like to see how `doGet` is instantiating and calling the `Analyze` class to have more clarity on its intended use. It's fine to call it with dummy data for now and include it in the PR. This way, Bless can update the dummy data later to real data once it's ready.
> 2. Do we need to close the Service after each request, or can we re-use it multiple times before closing it? If it's the latter, I would suggest to keep the service open for the duration of the `doGet` method call, so we don't need to re-create the service if we have multiple posts to process.
> 3. Try to sub out the service, as suggested in the pod chat, so tests can be kept while not calling the real endpoint.
> 4. The lint / formatting checks are broken, so make sure to address them as otherwise the PR cannot be merged.

Since the main functionality of the methods is to call the language API, I'm not sure if subbing out the service would make sense in this situation.

jbiral commented:
Currently, yes, but you could imagine it will do more processing of the sentiment score later down the road, in which case it'll be helpful to be able to stub the service.

The thing is we can't write a test as long as we're calling the real API. And that may be fine for this CL, but we will need to stub something at some point (either this class, or the service and suggested)

shivanighanta committed:
Added helper methods

shivanighanta committed:
fixed formatting

shivanighanta commented:
Ok, that makes sense. I'll do that in the next PR when I process the sentiment score further.

ginellegaisano commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -34,292 +33,211 @@
   /** Identifies entities in a given string. */
-  public static void analyzeEntitiesText(String text) throws Exception {
-
+  public void analyzeEntitiesText(String text) throws Exception {
```
Should these return Entity?

ginellegaisano commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -34,292 +33,211 @@
-          System.out.printf("Content: %s\n", mention.getText().getContent());
-          System.out.printf("Type: %s\n\n", mention.getType());
-        }
+  public void analyzeEntities(Document doc, LanguageServiceClient language) throws Exception {
```
I think you'll probably want this one to be private, so that callers need to use analyzeEntitiesFile or analyzeEntitiesText instead. Also with the rest of the helpers.

ginellegaisano commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -34,292 +33,211 @@
 
   /** Detects categories in a given string using the Language API. */
-  public static void classifyText(String text) throws Exception {
+  public void classifyText(String text) throws Exception {
```
Should this return ClassificationCategory? Same below.

jbiral commented on src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -32,9 +33,21 @@
  */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  private Analyze analyze = new Analyze();
```
nit: add an empty line before and after this variable (and let's make it final too).

ginellegaisano commented on src/main/java/com/google/sps/language/Analyze.java:
```diff
@@ -34,292 +33,211 @@
 
   /** Detects the entity sentiments in the given string using the Language API. */
-  public static void entitySentimentText(String text) throws Exception {
+  public void entitySentimentText(String text) throws Exception {
```
Should these return Entity? Same below.

shivanighanta replied:
I wanted to print out all the enitity attributes for this PR to check the natural languge API is returning all the right things. After checking how the methods interact with the youtube API I'll change the return values of the methods to what makes most sense.

jbiral commented on src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -32,9 +33,21 @@
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("text/html;");
     response.getWriter().println("<h1>Hello world!</h1>");
```
Could you update the response with a sentiment score, for example? I know it won't be final, but this way we can see a reflection of the sentiment API into the response

jbiral commented on src/test/java/com/google/sps/AnalyzeTest.java:
```diff
@@ -0,0 +1,100 @@
+import org.junit.Ignore;
+
+@RunWith(JUnit4.class)
+@Ignore
```
Please add a comment to explain why the test is ignored:

```java
@Ignore // Comment here
```

ginellegaisano commented on src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -12,14 +12,15 @@
 // limitations under the License.
 
-package com.google.sps.servlets;
+package com.google.sps;
```
OOC, why did this package name change?

shivanighanta replied:
I wasn't able to access the methods in the Analyze class unless they were in the same package.

jonaner reviewed (approved):
In addition to prior reviews, I think this looks good.

shivanighanta committed:
Fixed

shivanighanta committed:
merging

shivanighanta committed:
merge

shivanighanta committed:
fixed

shivanighanta committed:
formatting
